### PR TITLE
Relax assertions in sockaddr_storage_to_addr to match the documentation.

### LINKED ISSUE #1479: sockaddr_storage_to_addr is documented as being more lenient with the size than it is in practice
The documentation says

```
The len argument should be the number of bytes in the sockaddr_storage that are actually allocated and valid. It must be at least as large as all the useful parts of the structure.
```

This suggests that passing in a size larger than what is actually used (say, `size_of::<sockaddr_storage>()`) should be fine. But in the implementation, both `AF_INET` and `AF_INET6` assert that the size passed is exactly `size_of::<sockaddr_in>()`/`size_of::<sockaddr_in6>()`. Ideally IMO the implementation would match the documented behavior.

### PR DESCRIPTION
**Pull request by khuey**
Fixes #1479

### CODE REVIEW

**Review by asomers — CHANGES_REQUESTED** [review]
The tests you modified are now doing two completely different things.  Better to split them up, testing only the minimal amount of functionality in each test.

File: test/sys/test_socket.rs [hunk 1]
```diff
@@ -54,6 +78,35 @@ pub fn test_inetv6_addr_to_sock_addr() {
...
+    let actual = SocketAddr::V6(SocketAddrV6::new(ip, port, flowinfo, scope_id));
+    let addr = InetAddr::from_std(&actual);
+    let sockaddr = SockAddr::new_inet(addr);
+
+    let (storage, ffi_size) = {
+        let mut storage = MaybeUninit::<sockaddr_storage>::zeroed();
+        let storage_ptr = storage.as_mut_ptr().cast::<sockaddr_in6>();
+        let (ffi_ptr, ffi_size) = sockaddr.as_ffi_pair();
+        // XXX this does not pass.
+        // assert_eq!(mem::size_of::<sockaddr>(), ffi_size as usize);
```
Comment (asomers) [inline test/sys/test_socket.rs:97]: You should delete this comment

**Review by asomers — CHANGES_REQUESTED** [review]

File: test/sys/test_socket.rs (hunk 1 above)
Comment (khuey) [reply test/sys/test_socket.rs]: Done.

**Comment by asomers** [comment]
You'll need to rebase to fix the test failures.

**Comment by khuey** [comment]
Done.

**Comment by asomers** [comment]
Just one more thing: would you mind squashing your commits?

**Comment by khuey** [comment]
Done.

**Review by asomers — APPROVED** [review]
bors r+

**Comment by asomers** [comment]
bors retry

**Comment by khuey** [comment]
Thanks.

**Pull request merged**